src/main/java/net/glowstone/block/blocktype/BlockPiston.java

```diff
@@ -156,23 +181,104 @@ public void onRedstoneUpdate(GlowBlock me) {
+        GlowBlock block, BlockFace ignoredFace,
+        List<Block> blocksToMove, List<Block> blocksToBreak) {
+
+        boolean isPushing = (movementDirection == ignoredFace.getOppositeFace());
+        MaterialValueManager.ValueCollection materialValues = block.getMaterialValues();
+        PistonMoveBehavior moveBehavior =
+            isPushing ? materialValues.getPistonPushBehavior() : materialValues.getPistonPullBehavior();
+
+        if (block.isEmpty()) {
+            return true;
+        }
+
+        if (blocksToMove.size() >= PUSH_LIMIT) {
+            return false;
+        }
+
+        if (isPushing) {
+            switch (moveBehavior) {
+                case MOVE_STICKY:
+                case MOVE:
+                    blocksToMove.add(block);
+                    break;
+                case BREAK:
+                    blocksToBreak.add(block);
+                    return true;
+                case DONT_MOVE:
+                    return false;
+                default:
+                    return true;
             }
+        } else {
+            switch (moveBehavior) {
+                case MOVE_STICKY:
+                case MOVE:
+                    blocksToMove.add(block);
+                    break;
+                case BREAK:
+                case DONT_MOVE:
+                    return true;
+                default:
+                    return true;
+            }
+        }
 
-            setType(relativeBlock, block.getTypeId(), block.getData());
-            setType(block, 0, 0);
+        if (moveBehavior == PistonMoveBehavior.MOVE_STICKY) {
+            boolean allowMovement = true;
+            for (BlockFace face : ADJACENT_FACES) {
+                GlowBlock nextBlock = block.getRelative(face);
 
-            return;
+                if (nextBlock.getLocation().equals(piston.getLocation())) {
+                    continue;
+                }
+
+                if (face == ignoredFace || blocksToMove.contains(nextBlock)) {
+                    continue;
+                }
+
+                allowMovement = addBlock(piston, movementDirection,
+                    block.getRelative(face), face.getOppositeFace(), blocksToMove, blocksToBreak);
+
+                if (!allowMovement) {
+                    break;
+                }
+            }
+            return allowMovement;
+        } else if (movementDirection != ignoredFace) {
+            GlowBlock nextBlock = block.getRelative(movementDirection);
+            if (nextBlock.getLocation().equals(piston.getLocation())) {
+                return false;
+            }
+
+            return addBlock(piston, movementDirection,
+                nextBlock, movementDirection.getOppositeFace(), blocksToMove, blocksToBreak);
+        } else {
+            return true;
         }
+    }
 
-        // remove piston head after retracting
-        setType(me.getRelative(pistonBlockFace), 0, 0);
+    private void breakBlock(GlowBlock block) {
```

Comment:
This logic doesn't take into account stuff like the `doTileDrops` gamerule and breaking containers (dropping chest contents). This kind of logic is already implemented in the [DiggingMessage handler](https://github.com/GlowstoneMC/Glowstone/blob/669ad8d45de9bea6ebe53a54bd6906fe9b4c5c87/src/main/java/net/glowstone/net/handler/play/player/DiggingHandler.java#L172-L185). Could you somehow merge this code?

If not, please leave a TODO and open an issue to support this.